### PR TITLE
core: prevent user_ta resource to be unpaged

### PR DESCRIPTION
**Pull request by etienne-lms**
Function `is_user_ta_ctx()` is used by unpaged code. Prior this change,
the whole `user_ta_ops` structure fell into unpaged sections. This change
decreases the unpaged size by several tenths of kB.

Below are extracts from mem_usage.py output on qemu_armv8 with
`CFG_WITH_PAGER=y` and `CFG_TEE_CORE_DEBUG=n`. Check the size of sections
.text and .rodata.

- Before the change:
```
RAM Usage        0E100000 - 0E166000 size 00066000 408 KiB 102 pages
.text            0E100000 - 0E1151E8 size 000151E8  84 KiB
.rodata          0E1151E8 - 0E117BD8 size 000029F0  10 KiB
*hole*           0E117BD8 - 0E118000 size 00000428   1 KiB
.data            0E118000 - 0E11A280 size 00002280   8 KiB
.bss             0E11A280 - 0E11CBD8 size 00002958  10 KiB
.heap1           0E11CBD8 - 0E120000 size 00003428  13 KiB
.nozi            0E120000 - 0E12D300 size 0000D300  52 KiB
.heap2           0E12D300 - 0E13A000 size 0000CD00  51 KiB
.text_init       0E13A000 - 0E13EEE0 size 00004EE0  19 KiB
.rodata_init     0E13EEE0 - 0E13FD10 size 00000E30   3 KiB
.rodata_pageable 0E13FD10 - 0E144080 size 00004370  16 KiB
.text_pageable   0E144080 - 0E166000 size 00021F80 135 KiB
```
- After the change:
```
RAM Usage        0E100000 - 0E167000 size 00067000 412 KiB 103 pages
.text            0E100000 - 0E108E48 size 00008E48  35 KiB
.rodata          0E108E48 - 0E10A3F0 size 000015A8   5 KiB
*hole*           0E10A3F0 - 0E10B000 size 00000C10   3 KiB
.data            0E10B000 - 0E10D278 size 00002278   8 KiB
*hole*           0E10D278 - 0E10D280 size 00000008   0 KiB
.bss             0E10D280 - 0E10FBE8 size 00002968  10 KiB
.heap1           0E10FBE8 - 0E110000 size 00000418   1 KiB
.nozi            0E110000 - 0E11D300 size 0000D300  52 KiB
.heap2           0E11D300 - 0E12D000 size 0000FD00  63 KiB
.text_init       0E12D000 - 0E133460 size 00006460  25 KiB
.rodata_init     0E133460 - 0E1342D0 size 00000E70   3 KiB
.rodata_pageable 0E1342D0 - 0E139A50 size 00005780  21 KiB
.text_pageable   0E139A50 - 0E167000 size 0002D5B0 181 KiB
```

### CODE REVIEW

File: core/arch/arm/kernel/user_ta.c [hunk 1]
```diff
@@ -502,14 +501,24 @@ static const struct tee_ta_ops user_ta_ops __rodata_unpaged = {
 static SLIST_HEAD(uta_stores_head, user_ta_store_ops) uta_store_list =
 		SLIST_HEAD_INITIALIZER(uta_stores_head);
 
+static struct tee_ta_ops const *_user_ta_ops = &user_ta_ops;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/user_ta.c]: sorry, i pushed a test version. With such an init, this change is useless.

File: core/arch/arm/kernel/user_ta.c [hunk 2]
```diff
@@ -490,7 +489,7 @@ static uint32_t user_ta_get_instance_id(struct tee_ta_ctx *ctx)
 	return to_user_ta_ctx(ctx)->vm_info->asid;
 }
 
-static const struct tee_ta_ops user_ta_ops __rodata_unpaged = {
+static const struct tee_ta_ops user_ta_ops = {
 	.enter_open_session = user_ta_enter_open_session,
 	.enter_invoke_cmd = user_ta_enter_invoke_cmd,
 	.enter_close_session = user_ta_enter_close_session,
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/kernel/user_ta.c]: This the (right) change applied, there are still some user_ta functions in the unpaged `.text` section: `to_user_ta_ctx()`, `user_ta_dump_state()` and `is_user_ta_ctx()`.

**Comment by etienne-lms** [comment]
buggy init fixed.

File: core/arch/arm/kernel/user_ta.c [hunk 3]
```diff
@@ -490,7 +489,7 @@ static uint32_t user_ta_get_instance_id(struct tee_ta_ctx *ctx)
 	return to_user_ta_ctx(ctx)->vm_info->asid;
 }
 
-static const struct tee_ta_ops user_ta_ops __rodata_unpaged = {
+static const struct tee_ta_ops user_ta_ops = {
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/user_ta.c]: Please keep the `__rodata_unpaged`. It's not supposed to affect the dependency tree, but it guarantees that  `user_ta_ops` itself doesn't end up in the paged part.

Comment (etienne-lms) [reply core/arch/arm/kernel/user_ta.c]: ok

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
i will add a comment stating why we use the service init way to init the reference.

**Comment by etienne-lms** [comment]
squashed.
comment added in the code.
tag applied.

File: core/arch/arm/kernel/user_ta.c [hunk 4]
```diff
@@ -502,14 +501,28 @@ static const struct tee_ta_ops user_ta_ops __rodata_unpaged = {
 static SLIST_HEAD(uta_stores_head, user_ta_store_ops) uta_store_list =
 		SLIST_HEAD_INITIALIZER(uta_stores_head);
 
+/*
+ * Break unpaged attribute propagation to user_ta_ops structure
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/user_ta.c]: I think it would be a little more clear with "unpaged attribute dependency propagation"

Comment (etienne-lms) [reply core/arch/arm/kernel/user_ta.c]: ok, thanks.

Comment (etienne-lms) [reply core/arch/arm/kernel/user_ta.c]: ok, thanks

**Comment by etienne-lms** [comment]
updated

**Pull request merged**